aplib.net-demo/Assets/Scripts/Configuration.cs

```diff
@@ -0,0 +1,277 @@
+
+    /// <summary>
+    /// Set inputfield amount to variable.
+    /// Then call check to validate value.
+    /// </summary>
+    /// <param name="amount">Amount of melee enemies.</param>
+    public void SetValueEnemyMelee(string amount)
+    {
+        if (amount == "") amount = "0";
+        EnemyMeleeAmount = int.Parse(amount);
+        CheckAllValuesToBeValid();
+    }
+
+    /// <summary>
+    /// Set inputfield amount to variable.
+    /// Then call check to validate value.
+    /// </summary>
+    /// <param name="amount">Amount of ranged enemies.</param>
+    public void SetValueEnemyRanged(string amount)
+    {
+        if (amount == "") amount = "0";
+        EnemyRangedAmount = int.Parse(amount);
+        CheckAllValuesToBeValid();
+    }
+
+    /// <summary>
+    /// This is called when the grid values are changed.
+    /// This is also called when an individual component has changed.
+    /// It updates all other conponents.
+    /// </summary>
+    private void CheckAllValuesToBeValid()
+    {
+        // Check gridW and gridH
+        if (GridSizeX > MaxGridSize) GridSizeX = MaxGridSize;
+        else if (GridSizeX < MinGridSize) GridSizeX = MinGridSize;
+
+        if (GridSizeZ > MaxGridSize) GridSizeZ = MaxGridSize;
+        else if (GridSizeZ < MinGridSize) GridSizeZ = MinGridSize;
+
+        // Update value.
+        GridVolume = GridSizeX * GridSizeZ;
+
+        // Check room value.
+        if (RoomAmount > GridVolume * _maxPercentage) RoomAmount = Mathf.CeilToInt(GridVolume * _maxPercentage);
+        else if (RoomAmount < 0) RoomAmount = Mathf.CeilToInt(GridVolume * _minPercentage);
+
+        // Check ItemAmmoAmount value.
+        if (ItemAmmoAmount > GridVolume * _maxPercentage) ItemAmmoAmount = Mathf.CeilToInt(GridVolume * _maxPercentage);
+        else if (ItemAmmoAmount < 0) ItemAmmoAmount = Mathf.CeilToInt(GridVolume * _minPercentage);
+
+        // Check ItemHealthAmount value.
+        if (ItemHealthAmount > GridVolume * _maxPercentage) ItemHealthAmount = Mathf.CeilToInt(GridVolume * _maxPercentage);
+        else if (ItemHealthAmount < 0) ItemHealthAmount = Mathf.CeilToInt(GridVolume * _minPercentage);
+
+        // Check ItemRageAmount value.
+        if (ItemRageAmount > GridVolume * _maxPercentage) ItemRageAmount = Mathf.CeilToInt(GridVolume * _maxPercentage);
+        else if (ItemRageAmount < 0) ItemRageAmount = Mathf.CeilToInt(GridVolume * _minPercentage);
+
+        // Check EnemyMeleeAmount value.
+        if (EnemyMeleeAmount > GridVolume * _maxPercentage) EnemyMeleeAmount = Mathf.CeilToInt(GridVolume * _maxPercentage);
+        else if (EnemyMeleeAmount < 0) EnemyMeleeAmount = Mathf.CeilToInt(GridVolume * _minPercentage);
+
+        // Check EnemyRangedAmount value.
+        if (EnemyRangedAmount > GridVolume * _maxPercentage) EnemyRangedAmount = Mathf.CeilToInt(GridVolume * _maxPercentage);
+        else if (EnemyRangedAmount < 0) EnemyRangedAmount = Mathf.CeilToInt(GridVolume * _minPercentage);
```

Comment:
Lots of repeated code where the same calculation is constantly repeated, consider doing something like this
```csharp
    int minValue = Mathf.CeilToInt(gridVolume * _minPercentage);
    int maxValue = Mathf.CeilToInt(gridVolume * _maxPercentage);

    RoomAmount = Mathf.Max(minValue, Mathf.Min(maxValue, RoomAmount));
    ItemAmmoAmount = Mathf.Max(minValue, Mathf.Min(maxValue, ItemAmmoAmount));
    etc.
```

Reply:
It looks so much cleaner! Thanks for the tip :D